nmigen_soc/periph.py

```diff
@@ -1,24 +1,146 @@
+from collections.abc import Mapping
+
+from nmigen.utils import bits_for
+
 from .memory import MemoryMap
 from . import event
 
 
-__all__ = ["PeripheralInfo"]
+__all__ = ["ConstantValue", "ConstantBool", "ConstantInt", "ConstantMap", "PeripheralInfo"]
+
+
+class ConstantValue:
+    pass
+
+
+class ConstantBool(ConstantValue):
+    """Boolean constant.
+
+    Parameters
+    ----------
+    value : bool
+        Constant value.
+    """
+    def __init__(self, value):
+        if not isinstance(value, bool):
+            raise TypeError("Value must be a bool, not {!r}".format(value))
+        self._value = value
+
+    @property
+    def value(self):
+        return self._value
+
+    def __repr__(self):
+        return "ConstantBool({})".format(self.value)
+
+
+class ConstantInt(ConstantValue):
+    """Integer constant.
+
+    Parameters
+    ----------
+    value : int
+        Constant value.
+    width : int
+        Width in bits. Optional. ``bits_for(value)`` by default.
+    signed : bool
+        Signedness. Optional. ``value < 0`` by default.
+    """
+    def __init__(self, value, *, width=None, signed=None):
+        if not isinstance(value, int):
+            raise TypeError("Value must be an integer, not {!r}"
+                            .format(value))
+        self._value = value
+
+        if width is None:
+            width = bits_for(value)
+        if not isinstance(width, int):
+            raise TypeError("Width must be an integer, not {!r}"
+                            .format(width))
+        if width < bits_for(value):
+            raise ValueError("Width must be greater than or equal to the number of bits needed to"
+                             " represent {}".format(value))
+        self._width = width
+
+        if signed is None:
+            signed = value < 0
+        if not isinstance(signed, bool):
+            raise TypeError("Signedness must be a bool, not {!r}"
+                            .format(signed))
+        self._signed = signed
+
+    @property
+    def value(self):
+        return self._value
+
+    @property
+    def width(self):
+        return self._width
+
+    @property
+    def signed(self):
+        return self._signed
+
+    def __repr__(self):
+        return "ConstantInt({}, width={}, signed={})".format(self.value, self.width, self.signed)
+
+
+class ConstantMap(Mapping):
+    """Named constant map.
+
+    A read-only container for named constants. Keys are iterated in ascending order.
```

Comment:
I don't have a strong opinion on this, but I feel like insertion order would be more useful--this way the order of constants in the BSP matches the order of constants in peripheral code. This would matter for few people (most won't be looking at raw BSP sources), but it still seems nice; on the other hand, any deterministic iteration order is fine reproducibility-wise.